### PR TITLE
Cross chain ( L2 ➡️ L1 ) NFT metadata transfer

### PR DESCRIPTION
**Pull request by itzmeanjan**
## Changes

- This PR attempts to bring functionality to transfer NFT metadata from L2 to L1
- Specifically speaking, NFT uri to be emitted via event ( `TransferWithMetadata` ) in burn tx on L2
- After checkpointing, when proof to be submitted to RootChainManager, it needs to have `TransferWithMetadata` packed in proof
- URI to be set during minting token on L1, using `_setTokenURI`
- For that, L1 MintableERC721 token must implement `mint(address user, uint256 tokenId, string calldata uri) external`, which can only be invoked by PredicateRole owner i.e. MintableERC721Predicate.

### CODE REVIEW

**Review by imyourm8 — CHANGES_REQUESTED** [review]
According to 721 standard metadata also includes `name` and `symbol`, https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/IERC721Metadata.sol

I would suggest to burn token with all required fields by the standard.

**Comment by imyourm8** [comment]
I also think that what a lot of developers want to see - the ability to transfer some custom data associated with the token. Maybe we could include such data into the burning event, with minting using required fields and optional bytes. It however will force developers to prepare ERC721 contract and map it on matic side, so the encoding part will be correct, which is fine.

**Comment by itzmeanjan** [comment]
> According to 721 standard metadata also includes `name` and `symbol`, https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/IERC721Metadata.sol
> 
> I would suggest to burn token with all required fields by the standard.

Ok we need to make change in burn event signature.

**Comment by itzmeanjan** [comment]
> I also think that what a lot of developers want to see - the ability to transfer some custom data associated with the token. Maybe we could include such data into the burning event, with minting using required fields and optional bytes. It however will force developers to prepare ERC721 contract and map it on matic side, so the encoding part will be correct, which is fine.

Do you mean letting developers pass custom `bytes` data in `withdrawWithMetadata` method ?

**Comment by itzmeanjan** [comment]
Now developers can pass arbitrary metadata along with their token when exiting it from L2 to L1. It needs to be decoded & set properly in L1.

File: flat/ChildMintableERC721.sol [hunk 1]
```diff
@@ -2174,6 +2176,45 @@ contract ChildMintableERC721 is
...
+     * Should set `withdrawnTokens` mapping to `true` for the tokenId being withdrawn
+     * This transaction will be verified when exiting on root chain
+     *
+     * Before calling this function, you may want calling `encodeTokenMetadata`
+     * and get metadata to be transferred from L2 to L1 during exit
+     *
+     * @param tokenId tokenId to withdraw
+     * @param metaData Metadata which client wants to take along with it, generally return value of `encodeTokenMetadata`
+     */
+    function withdrawWithMetadata(uint256 tokenId, bytes calldata metaData) external {
```
Comment (imyourm8) [inline flat/ChildMintableERC721.sol:2191]: I think we shouldn't allow passing whatever data inside this function, it's better to keep such logic within the contract, so the developer has control over data passing back to L1.

Comment (imyourm8) [reply flat/ChildMintableERC721.sol:2191]: For example, there are some game-related data within the token, like speed/attack, the advanced user will just set any value he wants, ruining the game.

**Review by imyourm8 — CHANGES_REQUESTED** [review]

File: flat/ChildMintableERC721.sol (hunk 1 above)
Comment (itzmeanjan) [reply flat/ChildMintableERC721.sol:2191]: Yes that's correct. Putting metaData encoder function call inside `withdrawWithMetadata` solves our problem. 

What do you think ?

Comment (imyourm8) [reply flat/ChildMintableERC721.sol:2191]: Yes, remove it, and enforce metaData encoder

Comment (itzmeanjan) [reply flat/ChildMintableERC721.sol:2191]: > Yes, remove metadata argument, and enforce metaData encoder

Ok, I'll make that change.

Comment (itzmeanjan) [reply flat/ChildMintableERC721.sol]: Made changes.

**Review by imyourm8 — APPROVED** [review]

**Review by jdkanani — APPROVED** [review]

**Comment by chntr** [comment]
Does this mean that only withdrawWithMetadata can be defined rather than withdraw in the L2 contract?

I don't want to have an opportunity for users to withdraw without metadata being sent.

**Pull request merged**